src/services/transaction_watcher_service.ts

```diff
@@ -0,0 +1,116 @@
+        // TODO(oskar) - we are checking expiry inline, because TypeORM failed
+        // to work with LessThanOrEqual:
+        // const now = new Date();
+        // const transactionsToCheck = await this._transactionRepository.find({
+        //     where: [
+        //         {
+        //             status: TransactionStates.Mempool,
+        //             expiresAt: LessThanOrEqual(now),
+        //         },
+        //         {
+        //             status: TransactionStates.Submitted,
+        //             expiresAt: LessThanOrEqual(now),
+        //         },
+        //     ],
+        // });
+        const now = new Date();
+        const isExpired = tx.expectedAt <= now;
+        try {
+            const txInBlockchain = await this._web3Wrapper.getTransactionByHashAsync(tx.hash);
+            if (txInBlockchain !== undefined && txInBlockchain !== null && txInBlockchain.hash !== undefined) {
+                if (txInBlockchain.blockNumber !== null) {
+                    logger.info(
+                        `a transaction with a ${tx.status} status is already on the blockchain, updating status to confirmed`,
+                        { hash: txInBlockchain.hash },
+                    );
+
+                    tx.status = TransactionStates.Confirmed;
+                    return this._saveTransactionAsync(tx);
+                } else if (!isExpired && tx.status !== TransactionStates.Mempool) {
+                    logger.info(`a transaction with a ${tx.status} status is pending, updating status to mempool`, {
+                        hash: txInBlockchain.hash,
+                    });
+                    tx.status = TransactionStates.Mempool;
+                    return this._saveTransactionAsync(tx);
+                } else if (isExpired) {
+                    tx.status = TransactionStates.Stuck;
+                    return this._saveTransactionAsync(tx);
+                }
+            }
+        } catch (err) {
+            // TODO(oskar) this throws a type error when calling a tx that was
+            // already dropped. Can we catch this specific error?
+            // TypeError: Cannot read property 'blockNumber' of null
+            // at Object.unmarshalTransaction (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/src/marshaller.ts:85:32)
+            // at Web3Wrapper.<anonymous> (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/src/web3_wrapper.ts:253:40)
+            // at step (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/lib/src/web3_wrapper.js:43:23)
+            // at Object.next (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/lib/src/web3_wrapper.js:24:53)
+            // at fulfilled (/Users/overmorrow/projects/0x/0x-api/node_modules/@0x/web3-wrapper/lib/src/web3_wrapper.js:15:58)
+            // at process._tickCallback (internal/process/next_tick.js:68:7)
```

Comment:
This is a bug in Web3Wrapper. We should handle the case where `null` is returned instead of a txn, and we don't currently do that. For now, I would explicitly check for this error, and if it's found, do the logic below. If another error is thrown, we should log it and continue.

Reply:
I try to catch it with:
```
if (err instanceof TypeError)
```

Is that the correct way?